checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -0,0 +1,1710 @@
+      }
+    }
+  }
+
+  /**
+   * If the return type of the enclosing method is {@code @Owning}, transfer ownership of the return
+   * value and treat its obligations as satisfied by removing it from {@code obligations}.
+   *
+   * @param node a return node
+   * @param cfg the CFG of the enclosing method
+   * @param obligations the current set of tracked obligations. If ownership is transferred, it is
+   *     side-effected to remove the obligations of the returned value.
+   */
+  private void handleReturn(
+      ReturnNode node, ControlFlowGraph cfg, Set<ImmutableSet<LocalVarWithTree>> obligations) {
+    if (isTransferOwnershipAtReturn(cfg)) {
+      Node returnExpr = node.getResult();
+      Node temp = typeFactory.getTempVarForNode(returnExpr);
+      if (temp != null) {
+        returnExpr = temp;
+      }
+      if (returnExpr instanceof LocalVariableNode) {
+        removeObligationContainingVar(obligations, (LocalVariableNode) returnExpr);
+      }
+    }
+  }
+
+  /**
+   * Should we transfer ownership to the return type of the method corresponding to a CFG? Returns
+   * true when there is no {@link NotOwning} annotation on the return type.
+   *
+   * @param cfg the CFG of the method
+   * @return true iff we should transfer ownership to the return type of the method corresponding to
+   *     a CFG
+   */
+  private boolean isTransferOwnershipAtReturn(ControlFlowGraph cfg) {
+    if (checker.hasOption(MustCallChecker.NO_LIGHTWEIGHT_OWNERSHIP)) {
+      // Default to always transferring at return if not using LO, just like ECJ does.
+      return true;
+    }
+
+    UnderlyingAST underlyingAST = cfg.getUnderlyingAST();
+    if (underlyingAST instanceof UnderlyingAST.CFGMethod) {
+      // TODO: lambdas? I think in that case we'll return false, below.
+      MethodTree method = ((UnderlyingAST.CFGMethod) underlyingAST).getMethod();
+      ExecutableElement executableElement = TreeUtils.elementFromDeclaration(method);
+      return typeFactory.getDeclAnnotation(executableElement, NotOwning.class) == null;
+    }
+    return false;
+  }
+
+  /**
+   * Updates a set of obligations to account for an assignment.
+   *
+   * @param node the assignment
+   * @param obligations the set of obligations to update, which may be side-effected depending on
+   *     the assignment: assigning to an owning field might remove obligations, assigning to a new
+   *     local variable might modify an obligation (by increasing the size of its resource alias
+   *     set), etc.
+   */
+  private void handleAssignment(
+      AssignmentNode node, Set<ImmutableSet<LocalVarWithTree>> obligations) {
+    Node lhs = node.getTarget();
+    Element lhsElement = TreeUtils.elementFromTree(lhs.getTree());
+    // Use the temporary variable for the rhs if it exists.
+    Node rhs = removeCasts(node.getExpression());
+    LocalVariableNode tempVarForRhs = typeFactory.getTempVarForNode(rhs);
+    if (tempVarForRhs != null) {
+      rhs = tempVarForRhs;
+    }
```

Comment:
This pattern appears at least 3 times.  How about abstracting it out into a `getTempVarOrNode` method?